cmsplugin_feed_ai/feed.py

```diff
@@ -0,0 +1,190 @@
+        self.source_id = feed_obj.get("id", "")
+        self.text_content = feed_obj.get(
+            "message",
+            feed_obj.get("status_type", "").replace("_", " ").capitalize
+        )
+
+    @property
+    def permalink(self):
+        """
+        Permalink which points to the original feed object in external service.
+        """
+        return "%s/posts/%s" % (self.profile_url, self.source_id.split("_")[-1])
+
+
+def get_feed_settings(key=None, default=None):
+    feed_settings = getattr(settings, "AI_FEED_SETTINGS", {})
+    if key not in feed_settings:
+        warn("Missing %r setting in AI_FEED_SETTINGS" % key)
+        return default
+    return feed_settings.get(key, default)
+
+
+def get_feed_cache_key(service, profile_name, limit):
+    """Return a unique cache key for given service"""
+    return "cmsplugin_feed_ai_%s_%s_%s" % (service, profile_name, str(limit))
+
+
+def _fetch_twitter_profile_feed(profile_name, limit):
+    """
+    Fetch Twitter feed for the profile defined in settings.
+    Twython Docs: http://twython.readthedocs.io/en/latest/api.html?highlight=get_user_timeline#twython.Twython.get_user_timeline  # noqa
+    Twitter Docs: https://dev.twitter.com/rest/reference/get/statuses/user_timeline
+    """
+    conf = get_feed_settings("twitter")
+    api = Twython(
+        app_key=conf["app_key"],
+        app_secret=conf["access_token"],
+    )
+    return api.get_user_timeline(
+        screen_name=profile_name,
+        count=limit,
+    )
+
+
+def _fetch_facebook_profile_feed(profile_name, limit):
+    """
+    Fetch Facebook posts made by the profile defined in settings.
+    Docs: https://developers.facebook.com/docs/graph-api/reference/v2.7/user/feed
+    """
+    api = GraphAPI(access_token=get_feed_settings("facebook")["auth_token"])
+    wall = api.get_object(
+        "%s/posts" % (profile_name),
+        limit=limit,
+        fields="id,created_time,message,link,status_type",
+    )
+    return wall["data"]
+
+
+def _fetch_profile_feed(service, profile_name, limit):
+    """
+    Fetch profile feed from given service.
+    :ret type: Iterable of api dependant feed objects
+    """
+    if service == "twitter":
+        return _fetch_twitter_profile_feed(profile_name, limit)
+    elif service == "facebook":
+        return _fetch_facebook_profile_feed(profile_name, limit)
+
+
+def get_profile_feed(service, profile_name, limit=5):
+    """
+    Get display ready profile feed from given service.
+    This method uses cache to prevent excess calls to the service APIs.
+    :param service: Service identifier
+    :type service: Str
+    :ret type: List of FeedDisplayItems
+    """
+    key = get_feed_cache_key(service, profile_name, limit)
+    feed = cache.get(key)
+    if not feed:
+        feed = _fetch_profile_feed(service, profile_name, limit)
+        cache.set(key, feed, get_feed_settings("cache_time", DEFAULT_CACHE_TIME))
+    if service == "twitter":
```

Comment:
(and said set can then be used where we now hard-code `"twitter"` and `"facebook"` :)